File: Languages/cpp/low-level-design-lld/problems/rate_limiter.cpp
/**
 * LLD Problem: Rate Limiter
 *
 * Implement a rate limiter using the Token Bucket algorithm with a
 * sliding window approach.
 *
 * Token Bucket:
 * - Bucket has max capacity of tokens
 * - Tokens are added at a fixed refill rate
 * - Each request consumes one token
 * - Request denied if no tokens available
 *
 * Also includes a Sliding Window rate limiter for comparison.
 */
#include <iostream>
#include <deque>
#include <cassert>
#include <algorithm>
using namespace std;

// --- Token Bucket Rate Limiter ---
class TokenBucketLimiter {
public:
    TokenBucketLimiter(int maxTokens, double refillRate)
        : maxTokens_(maxTokens), tokens_(maxTokens),
          refillRate_(refillRate), lastRefill_(0.0) {}

    bool allowRequest(double timestamp) {
        refill(timestamp);
        if (tokens_ > 0) {
            tokens_--;
            return true;
        }
        return false;
    }

private:
    int maxTokens_;
    double tokens_;
    double refillRate_;  // tokens per second
    double lastRefill_;

    void refill(double timestamp) {
        double elapsed = timestamp - lastRefill_;
        double newTokens = elapsed * refillRate_;
        tokens_ = min((double)maxTokens_, tokens_ + newTokens);
        lastRefill_ = timestamp;
    }
};

// --- Sliding Window Rate Limiter ---
class SlidingWindowLimiter {
public:
    SlidingWindowLimiter(int maxRequests, double windowSeconds)
        : maxRequests_(maxRequests), windowSeconds_(windowSeconds) {}

    bool allowRequest(double timestamp) {
        // Remove expired timestamps
        while (!window_.empty() && window_.front() <= timestamp - windowSeconds_) {
            window_.pop_front();
        }
        if ((int)window_.size() < maxRequests_) {
            window_.push_back(timestamp);
            return true;
        }
        return false;
    }

private:
    int maxRequests_;
    double windowSeconds_;
    deque<double> window_;
};

int main() {
    // Test Token Bucket: 3 tokens, refill 1 token/sec
    {
        TokenBucketLimiter limiter(3, 1.0);

        assert(limiter.allowRequest(0.0) == true);   // 2 tokens left
        assert(limiter.allowRequest(0.1) == true);   // 1 token left
        assert(limiter.allowRequest(0.2) == true);   // 0 tokens left
        assert(limiter.allowRequest(0.3) == false);  // denied
        assert(limiter.allowRequest(0.5) == false);  // still not enough

        // After 1 second, 1 token refilled
        assert(limiter.allowRequest(1.3) == true);   // ~1 token refilled
        assert(limiter.allowRequest(1.4) == false);  // used it

        cout << "Token Bucket tests passed!" << endl;
    }

    // Test Sliding Window: 3 requests per 1-second window
    {
        SlidingWindowLimiter limiter(3, 1.0);

        assert(limiter.allowRequest(0.0) == true);
        assert(limiter.allowRequest(0.3) == true);
        assert(limiter.allowRequest(0.6) == true);
        assert(limiter.allowRequest(0.9) == false);  // window full

        // After window slides past first request
        assert(limiter.allowRequest(1.1) == true);   // 0.0 expired
        assert(limiter.allowRequest(1.2) == false);  // 0.3, 0.6, 1.1 in window

        cout << "Sliding Window tests passed!" << endl;
    }

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/low-level-design-lld/problems/rate_limiter.py
"""
Low-Level Design: Rate Limiter

Implement rate limiting using two strategies:
1. Token Bucket - tokens refill at fixed rate, each request consumes one
2. Sliding Window - track timestamps, reject if window is full

Both provide allowRequest(timestamp) -> bool interface.

Time: O(1) amortized for token bucket, O(1) amortized for sliding window
Space: O(1) for token bucket, O(max_requests) for sliding window
"""

from collections import deque


class TokenBucketLimiter:
    """Rate limiter using token bucket algorithm."""

    def __init__(self, max_tokens: int, refill_rate: float):
        """
        Args:
            max_tokens: Maximum tokens (burst capacity)
            refill_rate: Tokens added per second
        """
        self.max_tokens = max_tokens
        self.tokens = float(max_tokens)
        self.refill_rate = refill_rate
        self.last_refill = 0.0

    def allow_request(self, timestamp: float) -> bool:
        self._refill(timestamp)
        if self.tokens >= 1.0:
            self.tokens -= 1.0
            return True
        return False

    def _refill(self, timestamp: float) -> None:
        elapsed = timestamp - self.last_refill
        self.tokens = min(self.max_tokens, self.tokens + elapsed * self.refill_rate)
        self.last_refill = timestamp


class SlidingWindowLimiter:
    """Rate limiter using sliding window log."""

    def __init__(self, max_requests: int, window_seconds: float):
        self.max_requests = max_requests
        self.window_seconds = window_seconds
        self.timestamps: deque[float] = deque()

    def allow_request(self, timestamp: float) -> bool:
        # Evict expired entries
        while self.timestamps and self.timestamps[0] <= timestamp - self.window_seconds:
            self.timestamps.popleft()

        if len(self.timestamps) < self.max_requests:
            self.timestamps.append(timestamp)
            return True
        return False


if __name__ == "__main__":
    # Token Bucket: 3 tokens, refill 1/sec
    tb = TokenBucketLimiter(max_tokens=3, refill_rate=1.0)
    assert tb.allow_request(0.0) is True
    assert tb.allow_request(0.1) is True
    assert tb.allow_request(0.2) is True
    assert tb.allow_request(0.3) is False  # exhausted
    assert tb.allow_request(1.3) is True   # 1 token refilled
    assert tb.allow_request(1.4) is False  # used it
    print("Token Bucket tests passed!")

    # Sliding Window: 3 requests per 1 second
    sw = SlidingWindowLimiter(max_requests=3, window_seconds=1.0)
    assert sw.allow_request(0.0) is True
    assert sw.allow_request(0.3) is True
    assert sw.allow_request(0.6) is True
    assert sw.allow_request(0.9) is False  # window full
    assert sw.allow_request(1.1) is True   # 0.0 expired
    assert sw.allow_request(1.2) is False  # 0.3, 0.6, 1.1 still in window
    print("Sliding Window tests passed!")

    print("All tests passed!")
